File: backend/backend_test.js
#!/usr/bin/env node

const axios = require("axios");
const process = require("process");

class ColdEmailAPITester {
  constructor(baseUrl = "http://localhost:3000/api") {
    // original: https://coldemailai.preview.emergentagent.com/api
    this.baseUrl = baseUrl;
    this.token = null;
    this.userId = null;
    this.testsRun = 0;
    this.testsPassed = 0;
    this.testResults = [];
  }

  logTest(name, success, details = "") {
    this.testsRun += 1;
    if (success) {
      this.testsPassed += 1;
      console.log(`✅ ${name} - PASSED`);
    } else {
      console.log(`❌ ${name} - FAILED: ${details}`);
    }

    this.testResults.push({
      test: name,
      success,
      details,
      timestamp: new Date().toISOString(),
    });
  }

  async runTest(name, method, endpoint, expectedStatus, data = null, headers = null) {
    const url = `${this.baseUrl}/${endpoint}`;
    const testHeaders = {
      "Content-Type": "application/json",
    };

    if (this.token) {
      testHeaders["Authorization"] = `Bearer ${this.token}`;
    }
    if (headers) {
      Object.assign(testHeaders, headers);
    }

    console.log(`\n🔍 Testing ${name}...`);
    console.log(`   URL: ${url}`);

    try {
      let response;
      if (method === "GET") {
        response = await axios.get(url, { headers: testHeaders, timeout: 10000 });
      } else if (method === "POST") {
        response = await axios.post(url, data, { headers: testHeaders, timeout: 10000 });
      } else if (method === "DELETE") {
        response = await axios.delete(url, { headers: testHeaders, timeout: 10000 });
      } else {
        throw new Error(`Unsupported method: ${method}`);
      }

      const success = response.status === expectedStatus;
      let details = `Status: ${response.status}, Expected: ${expectedStatus}`;

      if (success) {
        try {
          const responseData = response.data;
          console.log(
            `   Response: ${JSON.stringify(responseData, null, 2).slice(0, 200)}...`
          );
          this.logTest(name, true, details);
          return [true, responseData];
        } catch {
          this.logTest(name, true, details);
          return [true, {}];
        }
      } else {
        let extra = "";
        try {
          extra = JSON.stringify(response.data);
        } catch {
          extra = String(response.data).slice(0, 100);
        }
        details += `, Error: ${extra}`;
        this.logTest(name, false, details);
        return [false, {}];
      }
    } catch (e) {
      const details = `Exception: ${e.message}`;
      this.logTest(name, false, details);
      return [false, {}];
    }
  }

  async testHealthCheck() {
    console.log("\n" + "=".repeat(50));
    console.log("TESTING HEALTH ENDPOINTS");
    console.log("=".repeat(50));

    await this.runTest("Health Check", "GET", "", 200);
    await this.runTest("API Health", "GET", "health", 200);
  }

  async testAuthFlow() {
    console.log("\n" + "=".repeat(50));
    console.log("TESTING AUTHENTICATION");
    console.log("=".repeat(50));

    const now = new Date();
    const timestamp = `${now.getHours()}${now.getMinutes()}${now.getSeconds()}`;
    const testEmail = `test_user_${timestamp}@example.com`;
    const testPassword = "TestPass123!";
    const testName = `Test User ${timestamp}`;

    const [success, response] = await this.runTest(
      "User Registration",
      "POST",
      "auth/register",
      200,
      {
        email: testEmail,
        password: testPassword,
        name: testName,
      }
    );

    if (success && response && response.access_token) {
      this.token = response.access_token;
      this.userId = response.user.id;
      console.log(`   Registered user: ${testEmail}`);
    } else {
      console.log("❌ Registration failed, cannot continue with auth tests");
      return false;
    }

    await this.runTest("Get Current User", "GET", "auth/me", 200);

    const [loginSuccess, loginResponse] = await this.runTest(
      "User Login",
      "POST",
      "auth/login",
      200,
      {
        email: testEmail,
        password: testPassword,
      }
    );

    if (loginSuccess && loginResponse && loginResponse.access_token) {
      this.token = loginResponse.access_token;
      console.log(`   Login successful for: ${testEmail}`);
    }

    return true;
  }

  async testCampaignFlow() {
    console.log("\n" + "=".repeat(50));
    console.log("TESTING CAMPAIGN WORKFLOW");
    console.log("=".repeat(50));

    if (!this.token) {
      console.log("❌ No auth token, skipping campaign tests");
      return false;
    }

    await this.runTest("Get Campaigns (Empty)", "GET", "campaigns", 200);

    const campaignBody = {
      name: "Test Campaign",
      work_description:
        "I'm a freelance web developer specializing in React and Node.js",
      email_limit: 10,
    };

    const [success, response] = await this.runTest(
      "Create Campaign",
      "POST",
      "campaigns",
      200,
      campaignBody
    );

    if (!success || !response || !response.id) {
      console.log("❌ Campaign creation failed, cannot continue");
      return false;
    }

    const campaignId = response.id;
    console.log(`   Created campaign: ${campaignId}`);

    await this.runTest(
      "Get Campaign Details",
      "GET",
      `campaigns/${campaignId}`,
      200
    );

    console.log("\n   Waiting for campaign processing...");
    await new Promise((r) => setTimeout(r, 3000));

    for (let i = 0; i < 3; i++) {
      const [ok, campaignData] = await this.runTest(
        `Campaign Progress Check ${i + 1}`,
        "GET",
        `campaigns/${campaignId}`,
        200
      );
      if (ok) {
        const status = campaignData.status || "unknown";
        const progress = campaignData.progress || 0;
        console.log(`   Status: ${status}, Progress: ${progress}%`);
      }
      await new Promise((r) => setTimeout(r, 2000));
    }

    await this.runTest("Get All Campaigns", "GET", "campaigns", 200);

    await this.runTest(
      "Delete Campaign",
      "DELETE",
      `campaigns/${campaignId}`,
      200
    );

    return true;
  }

  async testDashboardStats() {
    console.log("\n" + "=".repeat(50));
    console.log("TESTING DASHBOARD STATS");
    console.log("=".repeat(50));

    if (!this.token) {
      console.log("❌ No auth token, skipping dashboard tests");
      return false;
    }

    const [success, stats] = await this.runTest(
      "Get Dashboard Stats",
      "GET",
      "dashboard/stats",
      200
    );

    if (success && stats) {
      const expectedFields = [
        "total_campaigns",
        "active_campaigns",
        "total_emails_sent",
        "total_positive",
        "total_negative",
        "total_no_reply",
        "response_rate",
      ];
      for (const field of expectedFields) {
        if (field in stats) {
          console.log(`   ${field}: ${stats[field]}`);
        } else {
          console.log(`   ⚠️  Missing field: ${field}`);
        }
      }
    }

    return true;
  }

  async testErrorCases() {
    console.log("\n" + "=".repeat(50));
    console.log("TESTING ERROR CASES");
    console.log("=".repeat(50));

    await this.runTest(
      "Invalid Login",
      "POST",
      "auth/login",
      401,
      { email: "invalid@test.com", password: "wrongpass" }
    );

    if (this.token) {
      await this.runTest(
        "Duplicate Registration",
        "POST",
        "auth/register",
        400,
        {
          email: "test_user_123456@example.com",
          password: "TestPass123!",
          name: "Duplicate User",
        }
      );
    }

    const oldToken = this.token;
    this.token = "invalid_token";
    await this.runTest("Unauthorized Access", "GET", "campaigns", 401);
    this.token = oldToken;

    if (this.token) {
      await this.runTest(
        "Invalid Campaign Data",
        "POST",
        "campaigns",
        422,
        { name: "", work_description: "", email_limit: 0 }
      );
    }
  }

  async runAllTests() {
    console.log("🚀 Starting Cold Email AI Agent API Tests");
    console.log(`Testing against: ${this.baseUrl}`);
    console.log("=".repeat(60));

    const startTime = Date.now();

    await this.testHealthCheck();
    const authSuccess = await this.testAuthFlow();

    if (authSuccess) {
      await this.testDashboardStats();
      await this.testCampaignFlow();
    }

    await this.testErrorCases();

    const endTime = Date.now();
    const duration = (endTime - startTime) / 1000;

    console.log("\n" + "=".repeat(60));
    console.log("TEST SUMMARY");
    console.log("=".repeat(60));
    console.log(`Tests Run: ${this.testsRun}`);
    console.log(`Tests Passed: ${this.testsPassed}`);
    console.log(`Tests Failed: ${this.testsRun - this.testsPassed}`);
    console.log(
      `Success Rate: ${((this.testsPassed / this.testsRun) * 100).toFixed(1)}%`
    );
    console.log(`Duration: ${duration.toFixed(2)} seconds`);

    if (this.testsPassed === this.testsRun) {
      console.log("🎉 All tests passed!");
      return 0;
    } else {
      console.log("❌ Some tests failed!");
      return 1;
    }
  }
}

(async () => {
  const tester = new ColdEmailAPITester(); // change base URL here if needed
  const code = await tester.runAllTests();
  process.exit(code);
})();

File: backend_test.py
#!/usr/bin/env python3

import requests
import sys
import json
import time
from datetime import datetime

class ColdEmailAPITester:
    def __init__(self, base_url="https://coldemailai.preview.emergentagent.com/api"):
        self.base_url = base_url
        self.token = None
        self.user_id = None
        self.tests_run = 0
        self.tests_passed = 0
        self.test_results = []

    def log_test(self, name, success, details=""):
        """Log test result"""
        self.tests_run += 1
        if success:
            self.tests_passed += 1
            print(f"✅ {name} - PASSED")
        else:
            print(f"❌ {name} - FAILED: {details}")
        
        self.test_results.append({
            "test": name,
            "success": success,
            "details": details,
            "timestamp": datetime.now().isoformat()
        })

    def run_test(self, name, method, endpoint, expected_status, data=None, headers=None):
        """Run a single API test"""
        url = f"{self.base_url}/{endpoint}"
        test_headers = {'Content-Type': 'application/json'}
        
        if self.token:
            test_headers['Authorization'] = f'Bearer {self.token}'
        if headers:
            test_headers.update(headers)

        print(f"\n🔍 Testing {name}...")
        print(f"   URL: {url}")
        
        try:
            if method == 'GET':
                response = requests.get(url, headers=test_headers, timeout=10)
            elif method == 'POST':
                response = requests.post(url, json=data, headers=test_headers, timeout=10)
            elif method == 'DELETE':
                response = requests.delete(url, headers=test_headers, timeout=10)

            success = response.status_code == expected_status
            details = f"Status: {response.status_code}, Expected: {expected_status}"
            
            if success:
                try:
                    response_data = response.json()
                    print(f"   Response: {json.dumps(response_data, indent=2)[:200]}...")
                    self.log_test(name, True, details)
                    return True, response_data
                except:
                    self.log_test(name, True, details)
                    return True, {}
            else:
                try:
                    error_data = response.json()
                    details += f", Error: {error_data}"
                except:
                    details += f", Response: {response.text[:100]}"
                self.log_test(name, False, details)
                return False, {}

        except Exception as e:
            details = f"Exception: {str(e)}"
            self.log_test(name, False, details)
            return False, {}

    def test_health_check(self):
        """Test basic health endpoints"""
        print("\n" + "="*50)
        print("TESTING HEALTH ENDPOINTS")
        print("="*50)
        
        self.run_test("Health Check", "GET", "", 200)
        self.run_test("API Health", "GET", "health", 200)

    def test_auth_flow(self):
        """Test complete authentication flow"""
        print("\n" + "="*50)
        print("TESTING AUTHENTICATION")
        print("="*50)
        
        # Generate unique test user
        timestamp = datetime.now().strftime('%H%M%S')
        test_email = f"test_user_{timestamp}@example.com"
        test_password = "TestPass123!"
        test_name = f"Test User {timestamp}"

        # Test registration
        success, response = self.run_test(
            "User Registration",
            "POST",
            "auth/register",
            200,
            data={
                "email": test_email,
                "password": test_password,
                "name": test_name
            }
        )
        
        if success and 'access_token' in response:
            self.token = response['access_token']
            self.user_id = response['user']['id']
            print(f"   Registered user: {test_email}")
        else:
            print("❌ Registration failed, cannot continue with auth tests")
            return False

        # Test get current user
        self.run_test("Get Current User", "GET", "auth/me", 200)

        # Test login with same credentials
        success, response = self.run_test(
            "User Login",
            "POST",
            "auth/login",
            200,
            data={
                "email": test_email,
                "password": test_password
            }
        )
        
        if success and 'access_token' in response:
            self.token = response['access_token']  # Update token
            print(f"   Login successful for: {test_email}")
        
        return True

    def test_campaign_flow(self):
        """Test complete campaign workflow"""
        print("\n" + "="*50)
        print("TESTING CAMPAIGN WORKFLOW")
        print("="*50)
        
        if not self.token:
            print("❌ No auth token, skipping campaign tests")
            return False

        # Test get campaigns (empty initially)
        self.run_test("Get Campaigns (Empty)", "GET", "campaigns", 200)

        # Test create campaign
        campaign_data = {
            "name": "Test Campaign",
            "work_description": "I'm a freelance web developer specializing in React and Node.js",
            "email_limit": 10
        }
        
        success, response = self.run_test(
            "Create Campaign",
            "POST",
            "campaigns",
            200,
            data=campaign_data
        )
        
        if not success or 'id' not in response:
            print("❌ Campaign creation failed, cannot continue")
            return False
            
        campaign_id = response['id']
        print(f"   Created campaign: {campaign_id}")

        # Test get specific campaign
        self.run_test(
            "Get Campaign Details",
            "GET",
            f"campaigns/{campaign_id}",
            200
        )

        # Wait for campaign processing to start
        print("\n   Waiting for campaign processing...")
        time.sleep(3)

        # Check campaign progress multiple times
        for i in range(3):
            success, campaign_data = self.run_test(
                f"Campaign Progress Check {i+1}",
                "GET",
                f"campaigns/{campaign_id}",
                200
            )
            if success:
                status = campaign_data.get('status', 'unknown')
                progress = campaign_data.get('progress', 0)
                print(f"   Status: {status}, Progress: {progress}%")
            time.sleep(2)

        # Test get all campaigns
        self.run_test("Get All Campaigns", "GET", "campaigns", 200)

        # Test delete campaign
        self.run_test(
            "Delete Campaign",
            "DELETE",
            f"campaigns/{campaign_id}",
            200
        )

        return True

    def test_dashboard_stats(self):
        """Test dashboard statistics"""
        print("\n" + "="*50)
        print("TESTING DASHBOARD STATS")
        print("="*50)
        
        if not self.token:
            print("❌ No auth token, skipping dashboard tests")
            return False

        success, stats = self.run_test("Get Dashboard Stats", "GET", "dashboard/stats", 200)
        
        if success:
            expected_fields = ['total_campaigns', 'active_campaigns', 'total_emails_sent', 
                             'total_positive', 'total_negative', 'total_no_reply', 'response_rate']
            for field in expected_fields:
                if field in stats:
                    print(f"   {field}: {stats[field]}")
                else:
                    print(f"   ⚠️  Missing field: {field}")

    def test_error_cases(self):
        """Test error handling"""
        print("\n" + "="*50)
        print("TESTING ERROR CASES")
        print("="*50)
        
        # Test invalid login
        self.run_test(
            "Invalid Login",
            "POST",
            "auth/login",
            401,
            data={"email": "invalid@test.com", "password": "wrongpass"}
        )

        # Test duplicate registration
        if self.token:
            # Try to register with existing email
            self.run_test(
                "Duplicate Registration",
                "POST",
                "auth/register",
                400,
                data={
                    "email": "test_user_123456@example.com",  # Likely exists from previous test
                    "password": "TestPass123!",
                    "name": "Duplicate User"
                }
            )

        # Test unauthorized access
        old_token = self.token
        self.token = "invalid_token"
        self.run_test("Unauthorized Access", "GET", "campaigns", 401)
        self.token = old_token

        # Test invalid campaign creation
        if self.token:
            self.run_test(
                "Invalid Campaign Data",
                "POST",
                "campaigns",
                422,
                data={"name": "", "work_description": "", "email_limit": 0}
            )

    def run_all_tests(self):
        """Run complete test suite"""
        print("🚀 Starting Cold Email AI Agent API Tests")
        print(f"Testing against: {self.base_url}")
        print("="*60)
        
        start_time = time.time()
        
        # Run test suites
        self.test_health_check()
        auth_success = self.test_auth_flow()
        
        if auth_success:
            self.test_dashboard_stats()
            self.test_campaign_flow()
        
        self.test_error_cases()
        
        # Print summary
        end_time = time.time()
        duration = end_time - start_time
        
        print("\n" + "="*60)
        print("TEST SUMMARY")
        print("="*60)
        print(f"Tests Run: {self.tests_run}")
        print(f"Tests Passed: {self.tests_passed}")
        print(f"Tests Failed: {self.tests_run - self.tests_passed}")
        print(f"Success Rate: {(self.tests_passed/self.tests_run*100):.1f}%")
        print(f"Duration: {duration:.2f} seconds")
        
        if self.tests_passed == self.tests_run:
            print("🎉 All tests passed!")
            return 0
        else:
            print("❌ Some tests failed!")
            return 1

def main():
    tester = ColdEmailAPITester()
    return tester.run_all_tests()

if __name__ == "__main__":
    sys.exit(main())